### PR TITLE
feat(perpetual) integrate funding rate store into perpetual position manager

### LINKED ISSUE #2086: Add fundingRateIdentifier to construction params
**Why**

`FundingRateApplier` needs to request the funding rate from the Store for a specific `identifier`. This `identifier` should be whitelisted by the DVM. So, it should probably be set at construction time.

**How**

It's already implemented as `identifier` but this is currently described as the `priceIdentifier`, but this should actually reference a `fundingRateIdentifier`. Each Perpetual contract should reference 2 identifiers: one for the underlying price index that its synthetic tokens should be tracking, one for the funding rate.

**Resolution**

Better comments and tests.

**Difficulty Score [1-10]**

This is a measure of the amount of developer time this task will require to be completed. Note: this should include design and planning time in addition to implementation time. The scale is not linear:

- 3: ~1 hour

### LINKED ISSUE #2089: Integrate funding rate applier into perpetual position manager
**Why**

The perpetual position manager should update the funding rate on each interaction with the contract. Additionally, token debt should be scaled according to the funding rate in the case of an emergency shutdown.

**How**

Update PerpetualPositionManager.sol and add associated unit tests. If needed, update FundingRateApplier.

**Difficulty Score [1-10]**

This is a measure of the amount of developer time this task will require to be completed. Note: this should include design and planning time in addition to implementation time. The scale is not linear:

- 6: ~1 day

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

Add integration of the funding rate applier into the perpetual position manager.

Issue: #2089

**Summary**

- Adds in the `updateFundingRate` modifier on key functions in `PerpetualPositionManager`.
- Adds the `_getFundingRateAppliedTokenDebt` within the `settleEmergencyShutdown` function.
- Adds asscoaites tests to `PerpetualPositionManager`.
- Small modification to `FundingRateApplier` to deal with inheritance tree issue.

**Details**

On all `notEmergencyShutdown` functions the `updateFundingRate` modifier was added to update the funding rate. TRV was scaled according to the funding rate in an emergency shutdown.

The tests added validate two main things: 
- funding rate is correctly updated on each function call & that it is not updated when an emergency shutdown.
- `settleEmergencyShutdown` returns the correct amount of collateral, scaled by the cumulative funding rate multiplier.

Note that the implementation of `FundingRateApplier` was modified slightly. Specifically, the contract no longer inherits from `Testable` in the same way that the `FeePayer` does. This is due to an error in common inheritance levels between `FeePayer` and `FundingRateApplier` that generated the following error in contract compilation:
```
contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:27:1: DeclarationError: Base constructor arguments given twice.
contract PerpetualPositionManager is FeePayer, FundingRateApplier, AdministrateeInterface {
^ (Relevant source part starts here and spans across multiple lines).
contracts/financial-templates/common/FeePayer.sol:77:14: First constructor call is here:
    ) public Testable(_timerAddress) nonReentrant() {
             ^---------------------^
contracts/financial-templates/common/FundingRateApplier.sol:88:14: Second constructor call is here:
    ) public Testable(_timerAddress) nonReentrant() {
             ^---------------------^
```

This is as a result of the PerpetualPositionManager` both being a `FeePayer` and `FundingRateApplier` at the same level of the inheritance tree.

To overcome this error the `FundingRateApplier` now implements an instance of `Timer` to prevent it from needing to inherit from `Testable`.

**Issue(s)**

close #2089
close #2086

### CODE REVIEW

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol [hunk 1]
```diff
@@ -415,9 +420,9 @@ contract PerpetualLiquidatable is PerpetualPositionManager {
         // Note: all payouts are scaled by the unit collateral value so all payouts are charged the fees pro rata.
         FixedPoint.Unsigned memory feeAttenuation = _getFeeAdjustedCollateral(liquidation.rawUnitCollateral);
         FixedPoint.Unsigned memory settlementPrice = liquidation.settlementPrice;
-        FixedPoint.Unsigned memory tokenRedemptionValue = liquidation.tokensOutstanding.mul(settlementPrice).mul(
-            feeAttenuation
-        );
+        FixedPoint.Unsigned memory tokenRedemptionValue = _getFundingRateAppliedTokenDebt(liquidation.tokensOutstanding)
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol:423]: TODO: add a public method `getTokenDebt` that just returns the result of `_getFundingRateAppliedTokenDebt`

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+     *         INTERNAL FUNCTIONS           *
+     ****************************************/
+
+    function _getFundingRateStore() internal view returns (FundingRateStoreInterface) {
+        return FundingRateStoreInterface(fpFinder.getImplementationAddress("FundingRateStore"));
+    }
+
+    function _getLatestFundingRate() internal view returns (FixedPoint.Unsigned memory) {
+        FundingRateStoreInterface fundingRateStore = _getFundingRateStore();
+        return fundingRateStore.getLatestFundingRateForIdentifier(priceIdentifier);
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:147]: Will the funding rate have a different identifier than the underlying price?

For instance, there may be many tokens that track USDETH price. Each would likely have their own parameters, etc, and each would probably have its own funding rate identifier to determine each token's individual funding rate, right?

Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:147]: Yes I think the funding rate should have a diff identifier for the reason you identify. Chris and i discussed this and this [issue](https://github.com/UMAprotocol/protocol/issues/2086) came out of it

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol [hunk 3]
```diff
@@ -34,10 +34,10 @@ contract FundingRateApplier is Testable, Lockable {
     FinderInterface public fpFinder;
 
     // Last time the `cumulativeFundingRateMultiplier` was updated.
-    uint256 lastUpdateTime;
+    uint256 public lastUpdateTime;
 
     // Identifier in funding rate store to query for.
-    bytes32 identifier;
+    bytes32 public priceIdentifier;
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:40]: `priceIdentifier` is more consistent with how we previously defined this variable in the EMP.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol [hunk 4]
```diff
@@ -75,26 +76,23 @@ contract FundingRateApplier is Testable, Lockable {
...
         fpFinder = FinderInterface(_fpFinderAddress);
-        lastUpdateTime = getCurrentTime();
-        identifier = _identifier;
-    }
+        priceIdentifier = _priceIdentifier;
 
-    /****************************************
-     *         INTERNAL FUNCTIONS           *
-     ****************************************/
+        timer = Timer(_timerAddress);
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:93]: an instance of timer contract to remove the need to make `FundingRateApplier` a `Testable`.

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol [hunk 5]
```diff
@@ -58,6 +58,7 @@ contract PerpetualLiquidatable is PerpetualPositionManager {
         uint256 withdrawalLiveness;
         address collateralAddress;
         address finderAddress;
+        address fpFinderAddress;
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol:61]: only changes not to liquidatable are to pass in the params. No tests done on Liquidatable; this is left for a separate PR.

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 6]
```diff
@@ -117,7 +115,7 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
     }
 
     modifier notEmergencyShutdown() {
-        require(emergencyShutdownTimestamp == 0, "Contract emergency shutdown");
+        _notEmergencyShutdown();
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:118]: this was inconsistent. it should call the internal method `_notEmergencyShutdown`

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 7]
```diff
@@ -148,21 +146,26 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
...
         string memory _syntheticSymbol,
         address _tokenFactoryAddress,
         FixedPoint.Unsigned memory _minSponsorTokens,
         address _timerAddress,
         address _excessTokenBeneficiary
-    ) public FeePayer(_collateralAddress, _finderAddress, _timerAddress) nonReentrant() {
+    )
+        public
+        FeePayer(_collateralAddress, _finderAddress, _timerAddress)
+        FundingRateApplier(FixedPoint.fromUnscaledUint(1), _fpFinderAddress, _priceIdentifier, _timerAddress)
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:160]: for now, the initial funding rate is set to 1e18. should this be parameterized or do we always want it to start at 1?

File: packages/core/contracts/financial-templates/test/FundingRateApplierTest.sol [hunk 8]
```diff
@@ -10,9 +10,9 @@ contract FundingRateApplierTest is FundingRateApplier {
     constructor(
         FixedPoint.Unsigned memory _initialFundingRate,
         address _fpFinderAddress,
-        address _timerAddress,
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/test/FundingRateApplierTest.sol:13]: changed order of params to put the timer last, as is the convention in other contracts.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 3 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:40]: I think we should rename this to fundingRateIdentifier, #2086 , but can do in another PR

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol [hunk 9]
```diff
@@ -48,13 +48,15 @@ contract FundingRateApplier is Testable, Lockable {
...
      *                EVENTS                *
      ****************************************/
 
-    // TODO: Decide which params to emit in this event.
     event NewFundingRate(
         uint256 indexed newMultiplier,
+        uint256 lastUpdateTime,
         uint256 indexed updateTime,
         uint256 indexed paymentPeriod,
         uint256 latestFundingRate,
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:62]: nit: Maybe we make `latestFundingRate` indexed as it seems more informative.

File: packages/core/contracts/financial-templates/funding-rate-store/test/MockFundingRateStore.sol [hunk 10]
```diff
@@ -41,6 +41,9 @@ contract MockFundingRateStore is FundingRateStoreInterface, Testable {
         view
         returns (FixedPoint.Unsigned memory)
     {
+        if (fundingRates[identifier].length == 0) {
+            return FixedPoint.fromUnscaledUint(1);
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/funding-rate-store/test/MockFundingRateStore.sol:45]: I like this. Maybe on construction we just set the first one to 0 or set a default

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol (hunk 7 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:160]: I think this should be parameterized, for example we might want to start the first rate such that longs pay shorts by default

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 4 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:93]: Are there any consequences as a result of this? For example, what if a contract is FundingRateApplier but not FeePayer?

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 11]
```diff
@@ -338,6 +345,7 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
         external
         notEmergencyShutdown()
         fees()
+        updateFundingRate()
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:348]: Would be curious how much gas this adds to each contract action but can leave for another PR

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 12]
```diff
@@ -512,6 +523,8 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
      * @dev This burns all tokens from the caller of `tokenCurrency` and sends back the resolved settlement value of
      * `collateralCurrency`. Might not redeem the full proportional amount of collateral in order to account for
      * precision loss. This contract must be approved to spend `tokenCurrency` at least up to the caller's full balance.
+     * @dev Note that this function does not call the updateFundingRate modifier to update the funding rate as this
+     * function is only called after an emergency shutdown & there should be no funding rate updates after the shutdown.
      * @return amountWithdrawn The actual amount of collateral withdrawn.
      */
     function settleEmergencyShutdown()
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:530]: So just to clarify, no need to `updateFundingRate()` on this call because all collateral and debt should be frozen at this point?

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 13]
```diff
@@ -576,13 +595,18 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
     /****************************************
      *        GLOBAL STATE FUNCTIONS        *
      ****************************************/
+
+    function applyFundingRate() external notEmergencyShutdown() updateFundingRate() nonReentrant() {
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:599]: Smart to make this `notEmergencyShutdown()` if we're not applying `updateFundingRate()` on `settleEmergencyShutdown` +1

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js [hunk 14]
```diff
@@ -115,24 +118,33 @@ contract("PerpetualPositionManager", function(accounts) {
...
+      fpFinder.address, // _fpFinderAddress
       priceFeedIdentifier, // _priceFeedIdentifier
       syntheticName, // _syntheticName
       syntheticSymbol, // _syntheticSymbol
       tokenFactory.address, // _tokenFactoryAddress
       { rawValue: minSponsorTokens }, // _minSponsorTokens
       timer.address, // _timerAddress
       beneficiary, // _excessTokenBeneficiary
-      { from: contractDeployer }
+      { from: contractDeployer, gasLimit: 8000000000 }
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:147]: Does this fail without the `gasLimit`?

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js [hunk 15]
```diff
@@ -1016,6 +1029,242 @@ contract("PerpetualPositionManager", function(accounts) {
...
+    });
+
+    // requestTransferPosition
+    await timer.setCurrentTime((await timer.getCurrentTime()).add(toBN(1)).toString());
+    await positionManager.requestTransferPosition({ from: sponsor });
+    assert.equal((await positionManager.cumulativeFundingRateMultiplier()).toString(), toWei("0.95445")); // 0.945 * (1 + (1.01 - 1) * 1) = 0.95445
+
+    // requestTransferPosition. For this we will need to advance time by the withdrawlLiveness to pass the request withdraw.
+    // Set the funding rate to something more reasonable like 0.001% per second as 1.0001 as advancing over a long duration.
+    await mockFundingRateStore.setFundingRate(priceFeedIdentifier, await timer.getCurrentTime(), {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:1088]: WDYT about defining a helper function `setRateAndAdvanceTimer(rate, time)` that handles the `setFundingRate()` and `setCurrentTime()` repeated logic?

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js [hunk 16]
```diff
@@ -1249,7 +1498,7 @@ contract("PerpetualPositionManager", function(accounts) {
 
     // Create new oracle, replace it in the finder, and push a different price to it.
     const newMockOracle = await MockOracle.new(finder.address, timer.address);
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:1500]: Should we test precision loss on the multiplier in another PR?

**Review by nicholaspai — APPROVED** [review]
Really well made tests thanks, looks great to me I just have some nits about future issues/PR's

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 4 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:93]: I don't think there are any implications for this other than what you just described.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 2 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Yup. this is a good point and I agree that the identifier should be registered with the DVM as a separate identifier to the contracts `priceIdentifier` as UMA token holders will be required to vote on this. I have updated the PR accordingly.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 3 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:40]: have implemented this.

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol (hunk 7 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:160]: I think if we want to parameterize it then we should set the initial funding rate within the funding rate store. I've refactored the implementation of the `FundingRateApplier` to no longer take in any initial funding rate number; rather it calls `_getLatestFundingRate` within the constructor. If we want to seed a different value we should first set it in the funding rate store.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 9 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:62]: sure

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol (hunk 12 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:530]: correct

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js (hunk 14 above)
Comment (chrismaree) [reply packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:147]: works fine without it. have removed it.

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js (hunk 15 above)
Comment (chrismaree) [reply packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:1088]: we could refactor this but `setFundingRate` is only called at the top before the bulk of the tests and many functions require a slightly different setup such as withdraw requiring a new setup. IMHO it will be a bit messy to try to generalize this as 7 out of the 15 functions called in this test would require exception handling. can still refactor if you'd like.

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js (hunk 16 above)
Comment (chrismaree) [reply packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:1500]: ye, I think another PR. the precision loss I've seen in writing these tests is truncation at the 18th decimal in the cumulative funding rate, as expected.

**Comment by chrismaree** [comment]
Update: In order to accommodate comment https://github.com/UMAprotocol/protocol/pull/2084#discussion_r505969793 and issue https://github.com/UMAprotocol/protocol/issues/2086 I had to make some pretty big changes to the implementation of this PR. Specifically, adding in the `fundingRateIdentifier` resulted in us hitting the solidity stack limit, receiving the dreaded:
```
CompilerError: Stack too deep when compiling inline assembly: Variable headStart is 1 slot(s) too deep inside the stack.
```

To overcome this I made two small opinionated design changes to decrease the number of parameters required in constructing a `PerpetualLiquidatable` contract:
1) `PerpetualLiquidatable` & `PerpetualPositionManager` no longer take in the `fpFinder` address. Rather, the `PerpetualPositionManager` now asks the `Finder` for the Implementation address of a `"FinancialProductFinder"` within construction. This means that the finder will first require to register the `FinancialProductFinder` before deploying any perps. 

While I know we've wanted a clear separation between financial products (and their extensions, such as the `FundingRateStore`) I feel that putting the `fpFinder` as a lookup within the `Finder` actually makes sense given the overall UMA contract ecosystem. I am very curious for opinions on this change @nicholaspai @mrice32 @daywiss.

2) the `FundingRateApplier` no longer takes in an `initalFundingRate` param. Rather, it calls the `FundingRateStore` on construction to ask what the current funding rate is when seeding this initial value. If we want to seed the contract with a value other than 1e18 then we should simply set an initial funding rate in the `FundingRateStore` before deploying a new perp contract.

I know these are both opinionated changes and I am keen to hear some feedback from you guys. If we want to keep these two params we'll need to make some overall architecture changes to cut down the number of params that gets passed down the stack.

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 17]
```diff
@@ -149,13 +151,23 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
...
         string memory _syntheticSymbol,
         address _tokenFactoryAddress,
         FixedPoint.Unsigned memory _minSponsorTokens,
         address _timerAddress,
         address _excessTokenBeneficiary
-    ) public FeePayer(_collateralAddress, _finderAddress, _timerAddress) nonReentrant() {
+    )
+        public
+        FeePayer(_collateralAddress, _finderAddress, _timerAddress)
+        FundingRateApplier(
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:164]: this is the change where the `FundingRateApplier` gets the latest implementation address of the  `FinancialProductFinder`.

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol (hunk 7 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol]: >  If we want to seed a different value we should first set it in the funding rate store.
agreed

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js (hunk 15 above)
Comment (nicholaspai) [reply packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js]: Got it

File: packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js (hunk 16 above)
Comment (nicholaspai) [reply packages/core/test/financial-templates/perpetual-multiparty/PerpetualPositionManager.js:1511]: Ah so you've actually induced precision loss?

**Comment by nicholaspai** [comment]
> This means that the finder will first require to register the FinancialProductFinder before deploying any perps.

I really like this idea. It keeps the contract implementation simpler by making them always reference the same Finder, and I see no issues to having a Finder point to another Finder! It also is a good pattern if we want to create another set of contracts to point to like `predictionMarketsFinder` etc.

> Rather, it calls the FundingRateStore on construction to ask what the current funding rate is when seeding this initial value.
Agreed with this!

**Review by nicholaspai — APPROVED** [review]
LGTM

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 18]
```diff
@@ -149,13 +151,23 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
...
+            FinderInterface(_finderAddress).getImplementationAddress("FinancialProductFinder"),
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:165]: We should change this to `FundingRateStore` as discussed IRL
